lib/install/management/finder.go

```diff
@@ -311,62 +327,90 @@ func (d *Dispatcher) listResourcePools(path string) ([]*object.ResourcePool, err
+					continue
+				}
+				vchs = append(vchs, children...)
+			}
+		}
 	}
-
 	return vchs
 }
 
-// getVCHs will check vm with same name under this resource pool, to see if that's VCH vm, and it will also check children vApp, to see if that's a VCH.
-// eventually return all fond VCH VMs
-func (d *Dispatcher) getChildVCHs(pool *object.ResourcePool, searchVapp bool) ([]*vm.VirtualMachine, error) {
+// findVCHs finds any VCH in the specified pool.  If the compute resource has multiple pools, then look
+// for any VMs with the same name as the resource pool.  If the compute resource only had a single pool then
+// evaluate each VM in the pool for VCH metadata
+func (d *Dispatcher) findVCHs(pool *object.ResourcePool, multiPool bool) ([]*vm.VirtualMachine, error) {
 	defer trace.End(trace.Begin(pool.InventoryPath, d.op))
 
 	// check if pool itself contains VCH vm.
+	var vms []*vm.VirtualMachine
 	var vchs []*vm.VirtualMachine
-	poolName := pool.Name()
+	var err error
 	computeResource := compute.NewResourcePool(d.op, d.session, pool.Reference())
-	vmm, err := computeResource.GetChildVM(d.op, d.session, poolName)
-	if err != nil {
-		return nil, errors.Errorf("Failed to query children VM in resource pool %q: %s", pool.InventoryPath, err)
-	}
-	if vmm != nil {
-		vmm.InventoryPath = path.Join(pool.InventoryPath, poolName)
-		// #nosec: Errors unhandled.
-		if ok, _ := d.isVCH(vmm); ok {
-			d.op.Debugf("%q is VCH", vmm.InventoryPath)
-			vchs = append(vchs, vmm)
-		}
-	}
 
-	if !searchVapp {
-		return vchs, nil
-	}
+	// The compute resource had multiple pools, so the assumption is that any VCH that exists will be the same
+	// name as it's resource pool
+	if multiPool {
+		vm, err := computeResource.GetChildVM(d.op, d.session, pool.Name())
+		if err != nil {
+			return nil, errors.Errorf("Failed to query children VM in resource pool %q: %s", pool.InventoryPath, err)
+		}
+		if vm != nil {
+			vms = append(vms, vm)
+		}
 
-	vappPath := path.Join(pool.InventoryPath, "*")
-	vapps, err := d.session.Finder.VirtualAppList(d.op, vappPath)
-	if err != nil {
-		if _, ok := err.(*find.NotFoundError); ok {
-			return vchs, nil
+	} else {
+		// We only had one pool, so we'll look at all the VMs in that pool
+		vms, err = computeResource.GetChildrenVMs(d.op, d.session)
+		if err != nil {
+			return nil, errors.Errorf("Failed to query children VM in resource pool %q: %s", pool.InventoryPath, err)
 		}
-		d.op.Debugf("Failed to query vapp %q: %s", vappPath, err)
 	}
-	for _, vapp := range vapps {
-		childVCHs, err := d.getChildVCHs(vapp.ResourcePool, false)
-		if err != nil {
-			return nil, err
+	// iterate over VMs and determine if we've got any VCHs
+	for _, v := range vms {
+		// override the VM inventory path (which is folder based)
+		// for the resource pool path
+		v.InventoryPath = path.Join(pool.InventoryPath, v.Name())
+		// #nosec: Errors unhandled.
+		if ok, _ := d.isVCH(v); ok {
```

Comment:
This recursive query is O(n) for the number of VMs in the cluster in the robo case, or number of VMs under the compute resource (cluster or subpool with DRS), rather than related to number of pools or folders which should be a much smaller number.

If we're going to take this path I think we should be using a bulk query and update d.isVCH to take an array of possible VMs and return a list of those that are VCHs so this at least doesn't DoS the vSphere control plane.

Reply:
It will never be more than one in the DRS case (99%) ... only for Robo...we can change `isVCH`...it needs changed...but do we want to take the time for that now?